client/post-editor/editor-image-selector/preview.jsx

```diff
@@ -0,0 +1,223 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import { defer, isEqual, uniq } from 'lodash';
+import classNames from 'classnames';
+import Gridicon from 'gridicons';
+
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+import ImagePreloader from 'components/image-preloader';
+import { localize } from 'i18n-calypso';
+import MediaActions from 'lib/media/actions';
+import MediaStore from 'lib/media/store';
+import Spinner from 'components/spinner';
+import { url } from 'lib/media/utils';
+
+export class EditorImageSelectorPreview extends React.Component {
+	static displayName = 'EditorImageSelectorPreview';
+
+	static propTypes = {
+		siteId: PropTypes.number,
+		itemIds: PropTypes.array,
+		multiple: PropTypes.bool,
+		maxWidth: PropTypes.number,
+		onImageChange: PropTypes.func,
+		showEditIcon: PropTypes.bool,
+		translate: PropTypes.func,
+	};
+
+	state = {
+		images: [],
+	};
+
+	componentDidMount() {
+		this.fetchImages();
+		MediaStore.on( 'change', this.updateImageState );
+	}
+
+	componentDidUpdate( prevProps ) {
+		const { siteId, itemIds } = this.props;
+		if ( siteId !== prevProps.siteId || itemIds !== prevProps.itemIds ) {
+			this.fetchImages();
+		}
+	}
+
+	componentWillUnmount() {
+		MediaStore.off( 'change', this.updateImageState );
+	}
+
+	fetchImages = () => {
+		// We may not necessarily need to trigger a network request if we
+		// already have the data for the media item, so first update the state
+		this.updateImageState( () => {
+			if ( this.state.images ) {
+				return;
+			}
+
+			defer( () => {
+				this.props.itemIds.map( id => {
+					id = parseInt( id, 10 );
+					MediaActions.fetch( this.props.siteId, id );
+				} );
+			} );
+		} );
+	};
+
+	updateImageState = callback => {
+		if ( isEqual( this.state.images.map( image => image.ID ), this.props.itemIds ) ) {
+			return;
+		}
+
+		const images = uniq(
+			this.props.itemIds
+				.map( id => {
+					const media = MediaStore.get( this.props.siteId, id );
+					if ( ! media ) {
+						MediaActions.fetch( this.props.siteId, id );
+					}
+					return media;
+				} )
+				.filter( function( e ) {
+					return e;
+				} )
+		);
+
+		this.setState( { images }, () => {
+			if ( 'function' === typeof callback ) {
+				callback();
```

Comment:
Can you explain a bit why this particular callback on the setState is needed?

Reply:
I had handled this a bit incorrectly and udpated here - https://github.com/Automattic/wp-calypso/pull/26297/commits/cc8ed5029c0f70e3e30e9d515192a4a630969aa6

Images shouldn't be fetched every time we update the image state, only in the callback which checks if the props and image state are different and then fetches the images if not.  

We actually do need to run the updateImageState regularly to show transient images and fetched images asynchronously.